docs/0.3/workflow_specification.md

```diff
@@ -0,0 +1,179 @@
+# Workflow Specification
+
+<!--
+  Copyright (c) 2019-2021 Cargill Incorporated
+  Licensed under Creative Commons Attribution 4.0 International License
+  https://creativecommons.org/licenses/by/4.0/
+-->
+
+## Overview
+
+Grid Workflow enables users to model business processes as state transitions
+within smart contracts. Its design decouples the complex logic of business
+processes from smart contract logic, simplifying the creation of each.
+
+Workflow definitions are comprised of states, constraints, and permissions.
+Workflow states represent the stages of a business process. Constraints
+represent conditions that must be met to transition from one state to the next.
+Permission aliases define the actions that each party can take at each state in
+the workflow. Workflows contain one or more sub-workflows, which further help to
+encapsulate business process complexity.
+
+## Representation
+
+### Workflow
+
+`Workflow` consists of a list of `SubWorkflow`:
+```rust
+pub struct Workflow {
+    subworkflow: Vec<SubWorkflow>,
+}
+
+impl Workflow {
+    pub fn new(subworkflow: Vec<SubWorkflow>) -> Self {
+        Self { subworkflow }
+    }
+
+    pub fn subworkflow(&self, name: &str) -> Option<SubWorkflow> {
+        for sub_wf in &self.subworkflow {
+            if sub_wf.name() == name {
+                return Some(sub_wf.clone());
+            }
+        }
+
+        None
+    }
+}
+```
+
+### SubWorkflow
+
+`SubWorkflow` has a name, a list of `WorkflowState` and a list of starting
+states:
+
+```rust
+#[derive(Clone)]
+pub struct SubWorkflow {
+    name: String,
+    states: Vec<WorkflowState>,
+    starting_states: Vec<String>,
+}
+
+impl SubWorkflow {
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+
+    pub fn state(&self, name: &str) -> Option<WorkflowState> {
+        for state in &self.states {
+            if state.name() == name {
+                return Some(state.clone());
+            }
+        }
+
+        None
+    }
+
+    pub fn starting_states(&self) -> &[String] {
+        &self.starting_states
+    }
+}
+```
+
+### WorkflowState
+
+`WorkflowState` has a name, list of constraints, list of permission aliases,
+and list of transitions that can be made from that state. It has two methods:
+`can_transition` and `expand_permission`. `can_transition` returns true if an
+entity can execute a transition to a given state given its Pike permissions.
+`expand_permissions` returns a list of all permissions that are stored under a
+given `PermissionAlias`:
+
+```rust
+#[derive(Clone)]
+pub struct WorkflowState {
+    name: String,
+    constraints: Vec<String>,
+    permission_aliases: Vec<PermissionAlias>,
+    transitions: Vec<String>,
+}
+
+impl WorkflowState {
+    pub fn can_transition(&self, new_state: String,
+      pike_permissions: Vec<String>) -> bool {
+        if !self.transitions.contains(&new_state) {
```

Comment:
can you fix the formatting here to be : 
```
pub fn can_transition(
    &self,
    new_state: String,
    pike_permissions: Vec<String>,
) -> bool {...
```
to match what the rust compiler would fmt to (a little confusing to read otw)

Also, so we've added a check here for the workflow state's own name. Currently implemented in main. That change came after the RFC was merged, but I'm not sure if this document should stick to the RFC or current impl.

Reply:
I've updated it to match the implementation and cargo formatting (which puts the args on one line)